nncf/torch/knowledge_distillation_loss.py

```diff
@@ -0,0 +1,151 @@
+from copy import deepcopy
+from functools import reduce, partial
+
+from torch import nn
+import torch
+
+from nncf.common.schedulers import BaseCompressionScheduler
+from nncf.common.statistics import NNCFStatistics
+from nncf.torch.graph.transformations.layout import PTTransformationLayout
+from nncf import NNCFConfig
+from nncf.torch.nncf_network import NNCFNetwork
+from nncf.torch.compression_method_api import PTCompressionAlgorithmBuilder
+from nncf.torch.compression_method_api import PTCompressionAlgorithmController
+from nncf.api.compression import CompressionLoss, CompressionScheduler, CompressionStage
+from nncf.torch.algo_selector import COMPRESSION_ALGORITHMS
+from nncf.torch.compression_method_api import PTCompressionLoss
+from nncf.torch.nested_objects_traversal import nested_object_paths_generator
+
+
+class KnowledgeDistillationLoss(PTCompressionLoss):
+    def __init__(self, target_model: NNCFNetwork, original_model, kd_type):
+        super().__init__()
+        original_model.train()
+        device = next(target_model.parameters()).device
+        if kd_type == 'softmax':
+            def kd_loss_fn(ref_outputs, compressed_model_outputs):
+                return -(nn.functional.log_softmax(compressed_model_outputs, dim=1) *
+                         nn.functional.softmax(ref_outputs, dim=1)).mean() * (compressed_model_outputs.shape[1])
+        elif kd_type == 'mse':
+            def kd_loss_fn(ref_outputs, compressed_model_outputs):
+                mse = torch.nn.MSELoss()
+                return mse(ref_outputs, compressed_model_outputs)
+        else:
+            raise ValueError('Choose between mse/softmax options for Knowledge Distillation')
+        self._kd_loss_handler = target_model.create_knowledge_distillation_loss_handler(original_model, partial(
+            KnowledgeDistillationLoss._calculate,
+            device=device,
+            kd_loss_fn=kd_loss_fn))
+
+    @staticmethod
+    def _calculate(compressed_model_outputs, orig_model_outputs, device, kd_loss_fn):
+        """
+        Calculates knowledge distillation loss value from compressed_model_outputs and orig_model_outputs. First uses
+        nested_object_paths_generator to unpack input containers and numerate contents inside them.
+        Than checks compressed_model_outputs unpacked container for loss tensors (requires_grad=True)
+        and maps extracted structure of loss tensors to orig_model_outputs.
+        Finally computes knowledge distillation loss with extracted loss tensors.
+
+        :param compressed_model_outputs: Output tensors of compressed model can be any type of container that
+            iterates over internal items in the deterministic way.
+        :param orig_model_outputs: Output tensors of original model (used for distillation) can be any type of
+            container with deterministic detour.
```

Comment:
What did you mean by "detour" here? Reword this to be consistent with the docstring for the first param.